apps/src/code-studio/assets/SpriteUpload.jsx

```diff
@@ -92,73 +116,164 @@ export default class SpriteUpload extends React.Component {
+                largeImageError: TestStatus.pass,
+                smallImageWarning: TestStatus.pass
+              });
+            }
+          };
+        };
+        // File should be .png
+        const extensionIsInvalid = !file.name.toLowerCase().endsWith('.png');
+        // Filename cannot contain spaces or capital letters.
+        const filenameIsInvalid =
+          file.name.includes(' ') || /[A-Z]/.test(file.name);
+
+        if (extensionIsInvalid) {
+          this.setState({
+            uploadStatus: {
+              status: UploadStatus.badFile
+            },
+            extensionError: TestStatus.fail
+          });
+        } else {
+          this.setState({
+            extensionError: TestStatus.pass
+          });
         }
-      });
-    } else {
-      this.determineIfSpriteAlreadyExists(
-        spriteAvailability,
-        file.name.split('.')[0],
-        category
-      );
-    }
+        if (filenameIsInvalid) {
+          this.setState({
+            uploadStatus: {
+              status: UploadStatus.badFile
+            },
+            filenameError: TestStatus.fail
+          });
+        } else {
+          this.setState({
+            filenameError: TestStatus.pass
+          });
+        }
+        if (!extensionIsInvalid && !filenameIsInvalid) {
+          this.setState(
+            {
+              uploadStatus: {
+                status: UploadStatus.none
+              }
+            },
+            () => {
+              this.determineIfSpriteAlreadyExists(
+                spriteAvailability,
+                file.name.split('.')[0],
+                category
+              );
+            }
+          );
+        }
+      }
+    );
   };
 
   handleCategoryChange = event => {
+    event.persist();
     let {filename, spriteAvailability, fileData} = this.state;
-
-    this.setState({
-      category: event.target.value
-    });
-
-    if (fileData) {
-      this.determineIfSpriteAlreadyExists(
-        spriteAvailability,
-        filename.split('.')[0],
-        event.target.value
-      );
-    }
+    this.setState(
+      {
+        category: event.target.value,
+        metadata: ''
+      },
+      () => {
+        if (fileData) {
+          this.determineIfSpriteAlreadyExists(
+            spriteAvailability,
+            filename.split('.')[0],
+            event.target.value
+          );
+        }
+      }
+    );
   };
 
   handleAvailabilityChange = event => {
+    event.persist();
     let {filename, category, fileData} = this.state;
 
-    this.setState({
-      spriteAvailability: event.target.value,
-      category: ''
-    });
-
-    if (fileData) {
-      this.determineIfSpriteAlreadyExists(
-        event.target.value,
-        filename.split('.')[0],
-        category
-      );
-    }
+    this.setState(
+      {
+        spriteAvailability: event.target.value,
+        category: '',
+        metadata: ''
+      },
+      () => {
+        if (fileData) {
```

Comment:
To reduce duplication, I wonder if we can pull the 'if (fileData)' into the `this.determineIfSpriteAlreadyExists`. Thinking about it really quick, I feel like have a 'quick bailout' in `determindIfSpriteAlreadyExists` would make sense.